design/BufferOperations.md

```diff
@@ -0,0 +1,149 @@
+In particular a mapped buffer cannot be used in a `WebGPUCommandBuffer` given to `WebGPUQueue.submit`.
+The following must be true or a validation error occurs for mapWrite (resp. mapRead):
+
+ - The buffer must have been created with the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+ - `offset + size` must not overflow and be at most the size of the buffer
+ - Depending on the design of memory barriers, the buffer must be, or allowed to be in the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+
+The a mapped buffer can be unmapped with:
+
+```
+partial interface WebGPUBuffer {
+    void unmap();
+};
+```
+
+This operation invalidates all the `WebGPUMappedMemory` created from the buffer and puts the buffer in the unmapped state.
+The buffer must be in the mapped state otherwise a validation error occurs when `unmap` is called.
+
+### WebGPUMappedMemory
+
+`WebGPUMappedMemory` is an object representing a mapped region of a buffer that's both pollable and promise-like.
+
+It can be in one of three states: pending, available and invalidated.
+
+The pollable interface is:
+
+```
+partial interface WebGPUMappedMemory {
+    bool isPending();
+    ArrayBuffer getPointer();
+};
+```
+
+`isPending` return true if the object is in the pending state, false otherwise.
+`getPointer` returns an ArrayBuffer representing the buffer data if the object is in the available state, null otherwise.
+
+`WebGPUMappedMemory` is acts like a promise as in that it is then-able:
+
+```
+partial interface WebGPUMappedMemory {
+    void then(WebGPUMappedMemorySuccessCallback success,
+              optional WebGPUMappedMemoryErrorCallback error);
+};
+```
+
+This acts like a `Promise<ArrayBuffer>.then` that is resolved on the Javascript task boundary in which the implementation detects the GPU is done with the buffer.
+On that boundary:
+
+ - The `WebGPUMappedMemory` goes in the available state.
+ - If the `WebGPUMappedMemory` was created via `WebGPUBuffer.mapWrite`, its content is cleared to 0.
+ - `success` is called with the content of the memory as an argument.
+
+If `success` hasn't been called when the WebGPUMappedMemory gets invalidated (meaning the object is still in the pending state), `error` is called instead.
+
+When `WebGPUMappedMemory` goes from the available state to the invalidated state, the `ArrayBuffer` for its content gets neutered.
+
+## Immediate data upload
+
+Buffer mapping is the path with the least number of copies but it is often useful to upload data to a buffer *right now*, if only for debugging.
+A `WebGPUBuffer` operation is provided that takes an ArrayBuffer and copies its content at an offset in the buffer.
+
+```
+partial interface WebGPUBuffer {
+    void setSubData(ArrayBuffer data, u32 offset);
+}
+```
+
+This operation acts as if it was done after all previous "device-level" commands and before all subsequent "device-level" commands.
```

Comment:
How does it know what commands are "previous" and what are "subsequent"?

Reply:
Here previous and subsequent refer to the order of the calls in the Javascript program order.

Reply:
I think that's not very well-defined if we'll allow submission of commands from multiple workers.

IIRC, Vulkan does such "in-place" uploads via command buffers. Doing them the same way would make it a bit clearer what commands a before the upload and what are after (IIUC, not with multiple queues, though).

Reply:
Can we associate this with a particular point in the queue, perhaps by passing the queue as an argument, or making this a member function on the queue? It would make it clear what data would get clobbered and what wouldn't. Doing this would also solve the problem of what happens when this is called during the time when a buffer is mapped.

We've heard from Metal developers that submitting commands in an order different than they were recorded is an important use-case.